source/Calamari.Common/Features/StructuredVariables/StructuredConfigVariablesService.cs

```diff
@@ -78,26 +83,99 @@ List<string> MatchingFiles(RunningDeployment deployment, string target)
             return files;
         }
 
-        IFileFormatVariableReplacer GetReplacerForFile(string filePath, bool supportNonJsonReplacement)
+        IFileFormatVariableReplacer[] GetReplacersToTryForFile(string filePath, string? specifiedFileFormat, bool supportNonJsonReplacement)
         {
             if (!supportNonJsonReplacement)
             {
-                return jsonReplacer;
+                return new []
+                {
+                    jsonReplacer
+                };
             }
 
-            log.Info($"Feature toggle flag {ActionVariables.StructuredConfigurationFeatureFlag} detected. Trying replacers for all supported file formats.");
+            log.Info($"Feature toggle flag {ActionVariables.StructuredConfigurationFeatureFlag} detected. Considering replacers for all supported file formats.");
+
+            if (!string.IsNullOrWhiteSpace(specifiedFileFormat))
+            {
+                var specifiedReplacer = FindReplacerForFormat(specifiedFileFormat);
+
+                return new []
+                {
+                    specifiedReplacer
+                };
+            }
             
-            // TODO: when we support explicit specification of file formats, handle that here.
+            var guessBasedOnFilePath = FindBestReplacerForFilePath(filePath);
+            if (guessBasedOnFilePath != null)
+            {
+                return new []
+                {
+                    // For backwards compatibility, always try JSON first.
+                    jsonReplacer,
+                    guessBasedOnFilePath
+                };
+            }
 
-            var replacer = replacers.FirstOrDefault(r => r.IsBestReplacerForFileName(filePath));
-            if (replacer != null)
+            return new []
             {
-                log.Info($"The config file at '{filePath}' is being handled as format '{replacer.FileFormatName}' because of the filename.");
-                return replacer;
+                jsonReplacer
+            };
+        }
+
+        IFileFormatVariableReplacer FindReplacerForFormat(string specifiedFileFormat)
+        {
+            var specifiedReplacer = allReplacers
+                .FirstOrDefault(r => r.FileFormatName.Equals(specifiedFileFormat, StringComparison.OrdinalIgnoreCase));
+
+            if (specifiedReplacer == null)
+            {
+                var availableFileFormats = string.Join(", ", allReplacers.Select(r => r.FileFormatName));
+                var message = $"The file format specified ({specifiedFileFormat}) is invalid. "
+                              + $"The available options are: {availableFileFormats}";
+                throw new Exception(message);
             }
+
+            return specifiedReplacer;
+        }
+
+        IFileFormatVariableReplacer? FindBestReplacerForFilePath(string filePath)
+        {
+            return nonJsonReplacers.FirstOrDefault(r => r.IsBestReplacerForFileName(filePath));
```

Comment:
Rather than having the existance of nonJsonReplacers, maybe we could just do something like: 

`return allReplacers.FirstOrDefault(r => r.FileFormatName != StructuredConfigVariablesFileFormats.Json && r.IsBestReplacerForFileName(filePath))`? I was a little confused on the existance of the nonJsonReplacers to begin with.

Reply:
If we did that, we could also change the Method to be something like `FindBestNonJsonReplacerForFilePath`.